examples/mdetr/model.py

```diff
@@ -0,0 +1,370 @@
+            positive_map,
+        )
+
+        # Calculate MDETR loss with VQA losses
+        loss = self.loss(
+            model_output.pred_logits,
+            model_output.pred_boxes,
+            targets,
+            positive_map,
+            indices,
+            contrastive_query_embeddings,
+            contrastive_token_embeddings,
+            tokenized,
+            answer_preds,
+            answers,
+            answer_types,
+            weight_dict,
+        )
+
+        return MDETRVQAOutput(model_output, answer_preds, loss)
+
+
+def mdetr_for_vqa(
+    num_queries: int = 100,
+    num_classes: int = 255,
+    embedding_dim: int = 768,
+    transformer_d_model: int = 256,
+    transformer_num_heads: int = 8,
+    transformer_encoder_layers: int = 6,
+    transformer_decoder_layers: int = 6,
+    transformer_dim_feedforward: int = 2048,
+    transformer_dropout: float = 0.1,
+    return_intermediate_dec: bool = True,
+    matcher_cost_class: int = 1,
+    matcher_cost_bbox: int = 5,
+    matcher_cost_giou: int = 2,
+    no_object_weight: float = 0.1,
+    contrastive_dim: Optional[int] = None,
+    temperature: Optional[float] = None,
+):
+    hidden_dim = transformer_d_model
+    vqa_heads = mdetr_gqa_heads()
+    num_heads = len(vqa_heads.heads.keys())
+
+    model = mdetr_resnet101(
+        num_queries,
+        num_classes,
+        embedding_dim,
+        transformer_d_model,
+        transformer_num_heads,
+        transformer_encoder_layers,
+        transformer_decoder_layers,
+        transformer_dim_feedforward,
+        transformer_dropout,
+        return_intermediate_dec,
+        num_extra_query_embeddings=num_heads,
+    )
+    if contrastive_dim is not None:
+        contrastive_alignment_image_projection = nn.Linear(hidden_dim, contrastive_dim)
+        contrastive_alignment_text_projection = nn.Linear(hidden_dim, contrastive_dim)
+        contrastive_loss = partial(contrastive_alignment_loss, temperature=temperature)
+    else:
+        (
+            contrastive_alignment_image_projection,
+            contrastive_alignment_text_projection,
+            contrastive_loss,
+        ) = (None, None, None)
+
+    matcher = HungarianMatcher(matcher_cost_class, matcher_cost_bbox, matcher_cost_giou)
+
+    soft_token_loss = partial(
+        soft_token_prediction_loss, no_object_weight=no_object_weight
+    )
+
+    loss = MDETRLoss(
+        soft_token_loss=soft_token_loss,
+        box_losses=box_losses,
+        contrastive_alignment_loss=contrastive_loss,
+        vqa_losses=[masked_dict_cross_entropy, masked_dict_accuracy],
+    )
+
+    return MDETRForVQA(
+        model,
+        vqa_heads,
+        matcher,
+        loss,
+        contrastive_alignment_image_projection,
+        contrastive_alignment_text_projection,
+    )
+
+
+class MDETRPhraseGroundingOutput(NamedTuple):
+    model_output: MDETRModelOutput
+    loss: Dict[str, Tensor]
+
+
+class MDETRForPhraseGrounding(nn.Module):
```

Comment:
generally should considering splitting out into smaller diffs for different tasks